Tests/JSSpecCompiler/test-runner.cpp

```diff
@@ -0,0 +1,184 @@
+        auto result_content = read(result_path);
+
+        if (should_update_expectations && !result_content.is_error()) {
+            using namespace FileSystem;
+            MUST(copy_file_or_directory(expectation_path.string(), result_path.string(),
+                RecursionMode::Disallowed, LinkMode::Disallowed, AddDuplicateFileMarker::No));
+        }
+
+        auto expectation = read(expectation_path);
+
+        bool read_successfully = !(expectation.is_error() || result_content.is_error());
+        EXPECT(read_successfully);
+
+        if (read_successfully) {
+            bool are_equal = expectation.value() == result_content.value();
+            EXPECT(are_equal);
+
+            if (!are_equal) {
+                dbgln("Files {} and {} do not match!", expectation_path.string(), result_path.string());
+                auto maybe_diff = Diff::from_text(expectation.value(), result_content.value());
+                if (!maybe_diff.is_error()) {
+                    for (auto const& hunk : maybe_diff.value())
+                        MUST(Diff::write_unified(hunk, *out, Diff::ColorOutput::Yes));
+                }
+            }
+        }
+    }
+}
+
+TEST_CASE(test_regression)
+{
+    auto* source_dir_env = getenv("SERENITY_SOURCE_DIR");
+    auto* binary_dir_env = getenv("LAGOM_BINARY_DIR");
+
+    DeprecatedString source_dir;
+    DeprecatedString binary_dir;
+
+    if (source_dir_env == nullptr || binary_dir_env == nullptr) {
+        warnln("SERENITY_SOURCE_DIR or LAGOM_BINARY_DIR is not set. Assuming . for SERENITY_SOURCE_DIR and ./Build/lagom/bin for LAGOM_BINARY_DIR.");
+        source_dir = MUST(Core::System::getcwd());
+        binary_dir = LexicalPath::join(source_dir, "Build/lagom/bin"sv).string();
+    } else {
+        source_dir = source_dir_env;
+        binary_dir = binary_dir_env;
+    }
+
+    auto* update_expectations_env = getenv("JSSC_UPDATE_EXPECTATIONS");
+
+    bool should_update_expectations = false;
+    if (update_expectations_env != nullptr && strcmp(update_expectations_env, "1") == 0)
+        should_update_expectations = true;
+
+    auto temp_directory = MUST(FileSystem::TempFile::create_temp_directory());
+
+    auto path_to_compiler_binary = LexicalPath::join(binary_dir, compiler_binary);
+    auto path_to_captured_stderr = LexicalPath::join(temp_directory->path(), stderr_capture_filename);
+
+    for (auto const& test_description : regression_tests) {
+        for (auto const& source : test_description.sources) {
+            dbgln("Running {}...", source);
+
+            auto path_to_test = LexicalPath::join(source_dir, tests_root, source);
+            auto path_to_expectation = LexicalPath::join(source_dir, tests_root, DeprecatedString::formatted("{}.expectation", source));
+
+            auto arguments = build_command_line_arguments(path_to_compiler_binary, path_to_test, test_description);
+
+            // FIXME: We should invent better API to invoke a helper process.
```

Comment:
We have Core::Command, Core::Process::spawn(), and whatever `Utilities/run-tests.cpp` does. was this copied from one of those?

Reply:
Not really, I've written this myself :) `run-tests.cpp` seem to also use manual `posix_spawn`.
I guess I will then just add support for file actions to `Core::Process` and use it here and make a separate PR to port `run-tests.cpp` to it.